File: problems/0001-lexicographic-rank-sum-queries/solutions/python/solution.py
"""Problem: Lexicographic Rank Sum Queries.

Approach: suffix automaton with lexicographic subtree counts.
Time: O((|s| + total query length) * alphabet size).
Space: O(|s|).
"""

from __future__ import annotations

import json
import sys
from typing import Any


MODULUS = 1_000_000_007


def solve(input_data: dict[str, Any]) -> list[list[int]]:
    """Return the number and rank sum of substrings not greater than each query."""
    source = str(input_data["s"])
    queries = [str(query) for query in input_data["queries"]]

    maximum_states = 2 * len(source) + 1
    longest = [0] * maximum_states
    suffix_link = [-1] * maximum_states
    transitions: list[dict[str, int]] = [dict() for _ in range(maximum_states)]
    state_count = 1
    last = 0

    for character in source:
        current = state_count
        state_count += 1
        longest[current] = longest[last] + 1

        parent = last
        while parent != -1 and character not in transitions[parent]:
            transitions[parent][character] = current
            parent = suffix_link[parent]

        if parent == -1:
            suffix_link[current] = 0
        else:
            successor = transitions[parent][character]
            if longest[parent] + 1 == longest[successor]:
                suffix_link[current] = successor
            else:
                clone = state_count
                state_count += 1
                longest[clone] = longest[parent] + 1
                suffix_link[clone] = suffix_link[successor]
                transitions[clone] = transitions[successor].copy()
                while (
                    parent != -1
                    and transitions[parent].get(character) == successor
                ):
                    transitions[parent][character] = clone
                    parent = suffix_link[parent]
                suffix_link[successor] = clone
                suffix_link[current] = clone
        last = current

    order = sorted(range(state_count), key=longest.__getitem__)
    reachable_substrings = [0] * state_count
    for state in reversed(order):
        reachable_substrings[state] = sum(
            1 + reachable_substrings[next_state]
            for next_state in transitions[state].values()
        )

    ordered_transitions = [
        sorted(state_transitions.items())
        for state_transitions in transitions[:state_count]
    ]

    answers: list[list[int]] = []
    for query in queries:
        count = 0
        rank_sum = 0
        rank_offset = 0
        state = 0

        for character in query:
            for edge_character, next_state in ordered_transitions[state]:
                if edge_character >= character:
                    break
                block_size = 1 + reachable_substrings[next_state]
                rank_sum += block_size * rank_offset
                rank_sum += block_size * (block_size + 1) // 2
                rank_sum %= MODULUS
                count += block_size
                rank_offset += block_size

            next_state = transitions[state].get(character)
            if next_state is None:
                break

            state = next_state
            count += 1
            rank_offset += 1
            rank_sum = (rank_sum + rank_offset) % MODULUS

        answers.append([count, rank_sum])

    return answers


def main() -> None:
    json.dump(solve(json.load(sys.stdin)), sys.stdout)


if __name__ == "__main__":
    main()

File: problems/0001-lexicographic-rank-sum-queries/solutions/go/solution.go
// Problem: Lexicographic Rank Sum Queries
// Approach: Compact suffix automaton with lexicographic subtree counts.
// Time: O((|s| + total query length) * alphabet size).
// Space: O(|s| * alphabet size).
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

const modulus int64 = 1_000_000_007

type input struct {
	Source  string   `json:"s"`
	Queries []string `json:"queries"`
}

type state struct {
	next    [26]int32
	link    int32
	longest int32
}

type suffixAutomaton struct {
	states []state
	last   int32
}

func newSuffixAutomaton(capacity int) *suffixAutomaton {
	states := make([]state, 1, capacity)
	states[0].link = -1
	return &suffixAutomaton{states: states}
}

func (automaton *suffixAutomaton) extend(character byte) {
	edge := character - 'a'
	current := int32(len(automaton.states))
	automaton.states = append(automaton.states, state{
		link:    -1,
		longest: automaton.states[automaton.last].longest + 1,
	})

	parent := automaton.last
	for parent != -1 && automaton.states[parent].next[edge] == 0 {
		automaton.states[parent].next[edge] = current
		parent = automaton.states[parent].link
	}

	if parent == -1 {
		automaton.states[current].link = 0
	} else {
		successor := automaton.states[parent].next[edge]
		if automaton.states[parent].longest+1 == automaton.states[successor].longest {
			automaton.states[current].link = successor
		} else {
			clone := int32(len(automaton.states))
			cloneState := automaton.states[successor]
			cloneState.longest = automaton.states[parent].longest + 1
			automaton.states = append(automaton.states, cloneState)

			for parent != -1 && automaton.states[parent].next[edge] == successor {
				automaton.states[parent].next[edge] = clone
				parent = automaton.states[parent].link
			}
			automaton.states[successor].link = clone
			automaton.states[current].link = clone
		}
	}

	automaton.last = current
}

func buildSuffixAutomaton(text string) *suffixAutomaton {
	automaton := newSuffixAutomaton(2*len(text) + 1)
	for index := 0; index < len(text); index++ {
		automaton.extend(text[index])
	}
	return automaton
}

func (automaton *suffixAutomaton) reachableCounts(maximumLength int) []int64 {
	lengthCounts := make([]int, maximumLength+1)
	for index := range automaton.states {
		lengthCounts[automaton.states[index].longest]++
	}
	for length := 1; length <= maximumLength; length++ {
		lengthCounts[length] += lengthCounts[length-1]
	}

	order := make([]int32, len(automaton.states))
	for index := len(automaton.states) - 1; index >= 0; index-- {
		length := automaton.states[index].longest
		lengthCounts[length]--
		order[lengthCounts[length]] = int32(index)
	}

	reachable := make([]int64, len(automaton.states))
	for orderIndex := len(order) - 1; orderIndex >= 0; orderIndex-- {
		stateIndex := order[orderIndex]
		for _, nextState := range automaton.states[stateIndex].next {
			if nextState != 0 {
				reachable[stateIndex] += 1 + reachable[nextState]
			}
		}
	}
	return reachable
}

func triangularModulo(value int64) int64 {
	if value%2 == 0 {
		return ((value / 2) % modulus) * ((value + 1) % modulus) % modulus
	}
	return (value % modulus) * (((value + 1) / 2) % modulus) % modulus
}

func solve(data input) [][2]int64 {
	automaton := buildSuffixAutomaton(data.Source)
	reachable := automaton.reachableCounts(len(data.Source))
	answers := make([][2]int64, 0, len(data.Queries))

	for _, query := range data.Queries {
		var count int64
		var rankSum int64
		var rankOffset int64
		var stateIndex int32

		for characterIndex := 0; characterIndex < len(query); characterIndex++ {
			edge := query[characterIndex] - 'a'
			for smallerEdge := byte(0); smallerEdge < edge; smallerEdge++ {
				nextState := automaton.states[stateIndex].next[smallerEdge]
				if nextState == 0 {
					continue
				}
				blockSize := 1 + reachable[nextState]
				rankSum += (blockSize % modulus) * (rankOffset % modulus)
				rankSum += triangularModulo(blockSize)
				rankSum %= modulus
				count += blockSize
				rankOffset += blockSize
			}

			nextState := automaton.states[stateIndex].next[edge]
			if nextState == 0 {
				break
			}
			stateIndex = nextState
			count++
			rankOffset++
			rankSum = (rankSum + rankOffset) % modulus
		}

		answers = append(answers, [2]int64{count, rankSum})
	}
	return answers
}

func main() {
	var data input
	if err := json.NewDecoder(os.Stdin).Decode(&data); err != nil {
		fmt.Fprintln(os.Stderr, "invalid JSON input:", err)
		os.Exit(1)
	}
	if err := json.NewEncoder(os.Stdout).Encode(solve(data)); err != nil {
		fmt.Fprintln(os.Stderr, "failed to encode result:", err)
		os.Exit(1)
	}
}
